src/streaming-ingestion/streaming-ingestion-redshift.ts

```diff
@@ -0,0 +1,145 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
+ *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
+ *  and limitations under the License.
+ */
+
+import {
+  Stack,
+  NestedStack,
+  NestedStackProps,
+} from 'aws-cdk-lib';
+import { Role, IRole } from 'aws-cdk-lib/aws-iam';
+import { Construct } from 'constructs';
+import { StreamingIngestionProps } from './redshift/model';
+import { StreamingIngestionAssociateRole } from './redshift/streaming-ingestion-associate-role';
+import { StreamingIngestionSchemas } from './redshift/streaming-ingestion-schema';
+import { ExistingRedshiftServerlessProps, ProvisionedRedshiftProps } from '../analytics/private/model';
+import { RedshiftServerless } from '../analytics/private/redshift-serverless';
+import { addCfnNagForCustomResourceProvider, addCfnNagForLogRetention, addCfnNagToStack, ruleRolePolicyWithWildcardResources, ruleForLambdaVPCAndReservedConcurrentExecutions } from '../common/cfn-nag';
+import { SolutionInfo } from '../common/solution-info';
+
+type ExistingProvisionRedshiftProps = ProvisionedRedshiftProps & { dataAPIRoleArn: string};
+
+export interface StreamingIngestionRedshiftStackProps extends NestedStackProps {
+  readonly projectId: string;
+  readonly appIds: string;
+  readonly existingRedshiftServerlessProps?: ExistingRedshiftServerlessProps;
+  readonly existingProvisionedRedshiftProps?: ExistingProvisionRedshiftProps;
+  readonly streamingIngestionProps: StreamingIngestionProps;
+}
+
+export class StreamingIngestionRedshiftStack extends NestedStack {
+
+  readonly redshiftServerlessWorkgroup: RedshiftServerless | undefined;
+  readonly streamingIngestionAssociateRole: StreamingIngestionAssociateRole;
+  readonly streamingIngestionSchema: StreamingIngestionSchemas;
+  readonly redshiftDataAPIExecRole: IRole;
+  readonly streamingIngestionRole: IRole;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    props: StreamingIngestionRedshiftStackProps,
+  ) {
+    super(scope, id, props);
+
+    if (props.existingRedshiftServerlessProps && props.existingProvisionedRedshiftProps) {
+      throw new Error('Must specify ONLY one of existing Redshift Serverless or Provisioned Redshift.');
+    }
+
+    const featureName = `Streaming-Ingestion-${id}`;
+
+    this.templateOptions.description = `(${SolutionInfo.SOLUTION_ID}-sir) ${SolutionInfo.SOLUTION_NAME} - ${featureName} ${SolutionInfo.SOLUTION_VERSION_DETAIL}`;
+
+    let existingRedshiftServerlessProps: ExistingRedshiftServerlessProps | undefined = props.existingRedshiftServerlessProps;
+
+    const projectDatabaseName = props.projectId;
+    if (props.existingRedshiftServerlessProps) {
+      this.redshiftDataAPIExecRole = Role.fromRoleArn(this, 'StreamingIngestionDataExecRole',
+        props.existingRedshiftServerlessProps.dataAPIRoleArn, {
```

Comment:
We should move the `dataAPIRoleArn` for redshift props whatever it's servverless or provisioned redhsift.

Reply:
done